### ISSUE TITLE
nondeterministic behavior

### ISSUE THREAD

**Issue by aamster** [microsoft/LightGBM#2448]
I would expect lightgbm to produce deterministic results if the parameter `seed` is set. However I get differences of more than .002 ROC AUC on a validation set using multiple runs of the exact same code. 

I am using the following parameters:

lgb_params = {
    'objective':'binary',
    'boosting_type':'gbdt',
    'metric':'auc',
    'n_jobs': -1,
    'learning_rate':0.01,
    'num_leaves': 2**8,
    'max_depth':-1,
    'tree_learner':'serial',
    'colsample_bytree': 0.7,
    'subsample_freq':1,
    'subsample':0.7,
    'n_estimators':80000,
    'max_bin':255,
    'verbose':-1,
    'seed': 1234,
    'early_stopping_rounds':100, 
}

**Comment by StrikerRUS** [comment]
@aamster Please post the whole snippet. Also, attach the data you use, if it's not sensitive. Or can you reproduce this issue with random data?

**Comment by aamster** [comment]
Hi @StrikerRUS I found out that if I keep the data set fixed, then results are reproducible. However, if I regenerate the data, i.e generate all the features, which involve a lot of math (division, aggregations, etc) then I get different results each time. Is it possible that Lightgbm will choose different splits based on differences in precision for example, each time the features are generated? There is nothing random about the features that are generated.

**Comment by guolinke** [comment]
Did you change the order of row/column? If yes, it cannot assure the reproducible.

**Comment by aamster** [comment]
@guolinke yes I noticed that column order was changing. I believe this is the issue

**Comment by guolinke** [comment]
okay, as you set `colsample_bytree`, the sampled columns at each round will be different, for different column order.
